### PR TITLE
Addon-docs: Source loader library

### PR DESCRIPTION
**Pull request by shilman**
Issue: #7101 

## What I did

This is the source loader library by @libetl that replaces `addon-storysource`'s loader and extends it in various ways:

- storiesOf support
- module story format
- Typescript support
- Flow support

In addition, it can also embed the story source as a story parameter (which is probably how it should be implemented).

I'll leave the refactor of `addon-storysource` to @libetl for another PR. It might even be worth updating storysource to use the parameter and removing the complicated decorator/event.

This will be integrated into docs in a subsequent PR where the `Source` doc block consumes the aforementioned story parameter.

## How to test

```
yarn test
```

### CODE REVIEW

**Comment by igor-dv** [comment]
@shilman , I don't get why not to reference it in the storysource ? since it looks like a copy paste, hard to understand what exactly changed =) But It looks similar to what I remember it was =)

**Comment by shilman** [comment]
@igor-dv see my comment above:

> I'll leave the refactor of addon-storysource to @libetl for another PR. It might even be worth updating storysource to use the parameter and removing the complicated decorator/event.

**Comment by libetl** [comment]
@igor-dv you guessed it, storysource will just require it.
The idea behind it is to share the implementation

**Comment by igor-dv** [comment]
👍

File: lib/source-loader/src/abstract-syntax-tree/traverse-helpers.js [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      if (
+        node.type === 'ExportDefaultDeclaration' &&
+        node.declaration.type === 'ObjectExpression' &&
+        (node.declaration.properties || []).length
+      ) {
+        const parametersProperty = node.declaration.properties.find(
+          p => p.key.name === 'parameters' && p.value.type === 'ObjectExpression'
+        );
+
+        splicedSource =
```
Comment (shilman) [inline lib/source-loader/src/abstract-syntax-tree/traverse-helpers.js]: @libetl I think the error is here when there are no parameters attached to the default export

**Comment by shilman** [comment]
@libetl I think there's a bug when using the module format when there are no parameters on the default export. I tested this in the `release/docs-technical-preview` branch and it works on `cra-mdx` (a test project) but not in `official-storybook` (I tried commenting out various load lines, but all broke in the same place). Would you mind taking a look at this please?

**Comment by libetl** [comment]
fixed. But there seems to be something wrong after the load. all the stories are displaying errors. Can you help me Michael ?

**Comment by shilman** [comment]
@libetl Your fix worked but there is still a problem with some of the stories. In `PropsTable.stories.tsx`:

```
const { row: stringRow } = rowStories.string().props;
const { row: numberRow } = rowStories.number().props;
export const normal = () => <PropsTable rows={[stringRow, numberRow]} />;
```

When the source loader modifies the stories, the generated code assumes that the story functions are being called with context, and these are not.

Proposed solution:
1. I will modify the story format per https://github.com/storybookjs/storybook/issues/7071 and exprot the data directly
2. Later we could modify the source-loader to use injected story parameters instead of the event

**Comment by libetl** [comment]
Ok, tell me if I can help in any way

**Comment by shilman** [comment]
@libetl I was able to fix the problem as mentioned above.

There are two remaining problems, and I'm going to merge this and file a separate issue. Will update this as soon as I do.

**Review by ndelangen — APPROVED** [review]

File: .babelrc.js [hunk 2]
```diff
@@ -77,6 +78,7 @@ module.exports = {
         './lib/core/src/server',
         './lib/node-logger',
         './lib/codemod',
+        './lib/source-loader/src',
```
Comment (Hypnosphi) [inline .babelrc.js]: This directory includes client-side files like `preview.js`, we shouldn't apply node target to them

Comment (Hypnosphi) [reply .babelrc.js]: I'll extract the server part there

File: lib/source-loader/src/index.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { transform } from './build';
+
+export { STORY_EVENT_ID } from './events';
```
Comment (Hypnosphi) [inline lib/source-loader/src/index.js]: @shilman is that ewport used anywhere? I'm going to remove it from `index`

File: .babelrc.js (hunk 2 above)
Comment (Hypnosphi) [reply .babelrc.js]: #7510

**Pull request merged**